findmmrentries.go

```diff
@@ -0,0 +1,269 @@
+package veracity
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/datatrails/go-datatrails-common/logger"
+	"github.com/datatrails/go-datatrails-logverification/logverification/app"
+	"github.com/datatrails/go-datatrails-merklelog/massifs"
+	"github.com/datatrails/go-datatrails-merklelog/mmr"
+	"github.com/datatrails/go-datatrails-serialization/eventsv1"
+	veracityapp "github.com/datatrails/veracity/app"
+	"github.com/urfave/cli/v2"
+)
+
+/**
+ * find mmr Entries finds the mmr entry associated with given app data
+ */
+
+const (
+	appEntryFileFlagName = "app-entry-file"
+)
+
+// findMMREntries searchs the log of the given log tenant for matching mmrEntries given the app entries
+// and returns the leaf indexes of all the matches as well as the number of mmr entries considered
+func findMMREntries(
+	log logger.Logger,
+	massifReader MassifReader,
+	logTenant string,
+	massifStartIndex int64,
+	massifEndIndex int64,
+	massifHeight uint8,
+	appEntries ...[]byte,
+) ([]uint64, uint64, error) {
+
+	// find the starting leaf index by finding the number of leaf nodes in a full massif, of the given massif height,
+	//  then multiplying that by the number of massifs we are skipping over
+	leafIndex := uint64(massifStartIndex * int64(mmr.HeightIndexLeafCount(uint64(massifHeight-1))))
+
+	matchingLeafIndexes := []uint64{}
+	entriesConsidered := uint64(0)
+
+	massifIndex := massifStartIndex
+
+	// search all massifs from the starting index to the end index
+	for {
+
+		// check if we have reached the last massif we want to consider checking
+		// -1 means check until the last massif, so never break here if massifEndIndex == -1
+		if massifIndex > massifEndIndex && massifEndIndex != -1 {
+			break
+		}
+
+		massifContext, err := massifReader.GetMassif(context.Background(), logTenant, uint64(massifIndex))
+
+		// check if we have reached the last massif
+		if errors.Is(err, massifs.ErrMassifNotFound) {
+			break
+		}
+
+		if err != nil {
+
+			// check if we get an azblob error of blob not found
+			// this is also an indication we have reached the last massif
+			if strings.Contains(err.Error(), "BlobNotFound") {
+				break
+			}
+
+			return nil, 0, err
+		}
+
+		// get the mmrLeafEntries count based on the size
+		// NOTE: the leaf count and trie entry count are the same
+		// NOTE: the leaf index and trie index are equivilent.
+		mmrLeafEntries := massifContext.MassifLeafCount()
+
+		log.Debugf("checking %v trie entries in massif %v for matches", mmrLeafEntries, massifIndex)
+
+		// check each trie entry for matching trieKeys
+		for range mmrLeafEntries {
+
+			mmrIndex := mmr.MMRIndex(leafIndex)
```

Comment:
this smells a bit - its not obvious what the value of leafIndex is and not obvious that it is incremented at the end of the loop

if its difficult to do otherwise then add a comment

Reply:
will add a comment :)

Reply:
added comment!